### PR TITLE
EIP-1193 providers

### PR DESCRIPTION
**Pull request by avkos**
## Description

implementation of WS provider
#4656 
#5460

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

File: packages/web3-providers-ws/src/index.ts [hunk 1]
```diff
@@ -317,20 +315,19 @@ export default class WebSocketProvider<
...
+				this._eventEmitter.emit('message', response, undefined);
 				requestItem?.deferredPromise.reject(new ResponseError(response));
 			}
 
 			this._sentRequestsQueue.delete(requestId);
 		}
 	}
 
-	private _onConnect() {
-		this._reconnectAttempts = 0;
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/index.ts:332]: why this `_reconnectAttempts` is removed?

File: packages/web3-providers-ws/src/Eip1193Provider.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+			Web3APIMethod<API>,
+			ResponseType
+		>(
+			toPayload({
+				method: 'eth_chainId',
+				params: [],
+			}) as Web3APIPayload<API, Web3APIMethod<API>>,
+		);
+		return data?.result ?? '';
+		return '';
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/Eip1193Provider.ts]: This return statement can be removed

File: packages/web3-providers-ws/src/Eip1193Provider.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	private async _getAccounts(): Promise<HexString[]> {
+		const data = await (this as Web3BaseProvider<API>).request<Web3APIMethod<API>, HexString[]>(
+			toPayload({
+				method: 'eth_accounts',
+				params: [],
+			}) as Web3APIPayload<API, Web3APIMethod<API>>,
+		);
+		return data?.result ?? [];
+		return [];
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/Eip1193Provider.ts]: This return statement can be removed

File: packages/web3-providers-ws/src/Eip1193Provider.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+				}
+
+				this._eventEmitter.emit('connect', undefined, {
+					chainId: this._chainId,
+				});
+			})
+			.catch(err => {
+				// todo: add error handler
+				// eslint-disable-next-line no-console
+				console.log(err);
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/Eip1193Provider.ts]: for now:
```suggestion
				console.error(err);
```

File: packages/web3-providers-ws/src/Eip1193Provider.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+					)
+				) {
+					this._accounts = accounts;
+					this._onAccountsChanged();
+				}
+			})
+			.catch(err => {
+				// todo: add error handler
+				// eslint-disable-next-line no-console
+				console.log(err);
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/Eip1193Provider.ts]: for now
```suggestion
				console.error(err);
```

File: packages/web3-providers-ws/src/Eip1193Provider.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	HexString,
+	Web3APIMethod,
+	Web3APIPayload,
+	Web3APISpec,
+	Web3BaseProvider,
+} from 'web3-types';
+import { EventEmitter } from 'events';
+import { toPayload } from 'web3-utils';
+
+export default abstract class Eip1193Provider<
```
Comment (jdevcs) [inline packages/web3-providers-ws/src/Eip1193Provider.ts]: I think this abstract class should be placed in utils (or some other common) package so we can re use it in IPC provider + external users can use for making their providers eip1193

Comment (avkos) [reply packages/web3-providers-ws/src/Eip1193Provider.ts]: yes sure. It is just for draft pr. I will move it

File: packages/web3-providers-ws/src/Eip1193Provider.ts (hunk 2 above)
Comment (avkos) [reply packages/web3-providers-ws/src/Eip1193Provider.ts]: ok

File: packages/web3-providers-ws/src/Eip1193Provider.ts (hunk 3 above)
Comment (avkos) [reply packages/web3-providers-ws/src/Eip1193Provider.ts]: ok

File: packages/web3-providers-ws/src/Eip1193Provider.ts (hunk 4 above)
Comment (avkos) [reply packages/web3-providers-ws/src/Eip1193Provider.ts]: ok I will change to error

File: packages/web3-providers-ws/src/Eip1193Provider.ts (hunk 5 above)
Comment (avkos) [reply packages/web3-providers-ws/src/Eip1193Provider.ts]: ok I will change to error

File: packages/web3-providers-ws/src/index.ts (hunk 1 above)
Comment (avkos) [reply packages/web3-providers-ws/src/index.ts:332]: good catch. thanks

**Comment by jdevcs** [comment]
Thanks @avkos good work

**Review by nikoulai — APPROVED** [review]

**Review by jdevcs — COMMENTED** [review]
https://github.com/web3/web3.js/actions/runs/3700485520/jobs/6269015383#step:4:609. should be investigated as discussed.

**Comment by avkos** [comment]
I've created a separate task to investigate the stuck problem [#5730](https://github.com/web3/web3.js/issues/5730)

**Comment by jdevcs** [comment]
@avkos Please also add changelog in packages and also in root. Thanks

**Comment by jdevcs** [comment]
PR LGTM

**Review by jdevcs — APPROVED** [review]

**Pull request merged**